x/wasm/relay_pingpong_test.go

```diff
@@ -121,6 +120,189 @@ func TestPinPong(t *testing.T) {
 
 }
 
+func TestPinPongWithAppLevelError(t *testing.T) {
+	var (
+		coordinator = ibc_testing.NewCoordinator(t, 2)
+		chainA      = coordinator.GetChain(ibc_testing.GetChainID(0))
+		chainB      = coordinator.GetChain(ibc_testing.GetChainID(1))
+	)
+	_ = chainB.NewRandomContractInstance() // skip 1 id
+	var (
+		pingContractAddr = chainA.NewRandomContractInstance()
+		pongContractAddr = chainB.NewRandomContractInstance()
+	)
+	require.NotEqual(t, pingContractAddr, pongContractAddr)
+
+	pingContract := &player{t: t, actor: ping, chain: chainA, contractAddr: pingContractAddr}
+	pongContract := &player{t: t, actor: pong, chain: chainB, contractAddr: pongContractAddr}
+
+	wasmkeeper.MockContracts[pingContractAddr.String()] = pingContract
+	wasmkeeper.MockContracts[pongContractAddr.String()] = pongContract
+
+	var (
+		sourcePortID       = wasmkeeper.PortIDForContract(pingContractAddr)
+		counterpartyPortID = wasmkeeper.PortIDForContract(pongContractAddr)
+	)
+	clientA, clientB, connA, connB := coordinator.SetupClientConnections(chainA, chainB, clientexported.Tendermint)
+	connA.NextChannelVersion = ping
+	connB.NextChannelVersion = pong
+
+	channelA, channelB := coordinator.CreateChannel(chainA, chainB, connA, connB, sourcePortID, counterpartyPortID, channeltypes.UNORDERED)
+	var err error
+
+	const startValue uint64 = 100
+	const rounds = 3
+	s := startGame{
+		ChannelID: channelA.ID,
+		Value:     startValue,
+		MaxValue:  rounds - 1,
+	}
+	startMsg := &wasm.MsgExecuteContract{
+		Sender:   chainA.SenderAccount.GetAddress(),
+		Contract: pingContractAddr,
+		Msg:      s.GetBytes(),
+	}
+	// send from chainA to chainB
+	err = coordinator.SendMsgs(chainA, chainB, clientB, startMsg)
+	require.NoError(t, err)
+
+	t.Log("Duplicate messages are due to check/deliver tx calls")
+
+	var (
+		activePlayer  = ping
+		pingBallValue = startValue
+	)
+	for i := 1; i <= rounds-1; i++ { // play some rounds before reaching max value
+		t.Logf("++ round: %d\n", i)
+		ball := NewHit(activePlayer, pingBallValue)
+
+		seq := uint64(i)
+		pkg := channeltypes.NewPacket(ball.GetBytes(), seq, channelA.PortID, channelA.ID, channelB.PortID, channelB.ID, doNotTimeout, 0)
+		ack := ball.BuildAck()
```

Comment:
I just noticed this. The NewPacket creates and sends a packet, right, but it doesn't return the ack bytes?
You manually create the ack in the test code (not from the contract).

Is this due to unclear APIs? Is it not possible to read the packet/ack from the chain?

Reply:
`channeltypes.NewPacket` does not sent a packet here. It only creates an object that is used below with the coordinator to be passed to the other chain (B). This object creation would not be necessary in a full relayer demo. See my other comment.